### ISSUE TITLE
perldoc -f splice examples uses disfavored &foo(...) sub call

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#13321]
Migrated from [rt.perl.org#120051](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=120051) (status was 'resolved')

Searchable as RT120051$

**Comment by p5pRT** [comment]
### From revhippie@gmail.com

#### Created by revhippie@gmail.com

This is a bug report for perl from revhippie@&#8203;gmail\.com\,
generated with the help of perlbug 1\.39 running under perl 5\.16\.3\.


\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
'perldoc \-f splice' gives some example code&#8203;:

&nbsp;    sub aeq \{  \# compare two list values
&nbsp;        my\(@&#8203;a\) = splice\(@&#8203;\_\,0\,shift\);
&nbsp;        my\(@&#8203;b\) = splice\(@&#8203;\_\,0\,shift\);
&nbsp;        return 0 unless @&#8203;a == @&#8203;b;  \# same len?
&nbsp;        while \(@&#8203;a\) \{
&nbsp;            return 0 if pop\(@&#8203;a\) ne pop\(@&#8203;b\);
&nbsp;        \}
&nbsp;        return 1;
&nbsp;    \}
&nbsp;    if \(&aeq\($len\,@&#8203;foo\[1\.\.$len\]\,0\+@&#8203;bar\,@&#8203;bar\)\) \{ \.\.\. \}

That looks like an unnecessary use of the ampersand\.  Mildly less\-confusing
replacement&#8203;:

&nbsp;    if \(aeq\($len\,@&#8203;foo\[1\.\.$len\]\,0\+@&#8203;bar\,@&#8203;bar\)\) \{ \.\.\. \}

Although\, to be honest\, that seems like a terribly convoluted example\.

<details><summary>Perl Info</summary>

```

Flags:
    category=docs
    severity=low

Site configuration information for perl 5.16.3:

Configured by strawberry-perl at Tue Mar 12 12:12:58 2013.

Summary of my perl5 (revision 5 version 16 subversion 3) configuration:

  Platform:
    osname=MSWin32, osvers=4.0, archname=MSWin32-x64-multi-thread
    uname='Win32 strawberry-perl 5.16.3.1 #1 Tue Mar 12 12:12:07 2013 x64'
    config_args='undef'
    hint=recommended, useposix=true, d_sigaction=undef
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags =' -s -O2 -DWIN32 -DWIN64 -DCONSERVATIVE
 -DPERL_TEXTMODE_SCRIPTS -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS
-fno-strict-aliasing -mms-bitfields',
    optimize='-s -O2',
    cppflags='-DWIN32'
    ccversion='', gccversion='4.6.3', gccosandvers=''
    intsize=4, longsize=4, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long long', ivsize=8, nvtype='double', nvsize=8, Off_t='long
long', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='g++', ldflags ='-s -L"C:\strawberry\perl\lib\CORE"
-L"C:\strawberry\c\lib"'
    libpth=C:\strawberry\c\lib C:\strawberry\c\x86_64-w64-mingw32\lib
    libs=-lmoldname -lkernel32 -luser32 -lgdi32 -lwinspool -lcomdlg32
-ladvapi32 -lshell32 -lole32 -loleaut32 -lnetapi32 -luuid -lws2_32 -lmpr
-lwinmm -lversion -lodbc32 -lodbccp32 -lcomctl32
    perllibs=-lmoldname -lkernel32 -luser32 -lgdi32 -lwinspool -lcomdlg32
-ladvapi32 -lshell32 -lole32 -loleaut32 -lnetapi32 -luuid -lws2_32 -lmpr
-lwinmm -lversion -lodbc32 -lodbccp32 -lcomctl32
    libc=, so=dll, useshrplib=true, libperl=libperl516.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_win32.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags='-mdll -s -L"C:\strawberry\perl\lib\CORE"
-L"C:\strawberry\c\lib"'

Locally applied patches:



@INC for perl 5.16.3:
    C:/strawberry/perl/site/lib/MSWin32-x64-multi-thread
    C:/strawberry/perl/site/lib
    C:/strawberry/perl/vendor/lib
    C:/strawberry/perl/lib
    .


Environment for perl 5.16.3:
    HOME (unset)
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=C:\Program
Files\ImageMagick-6.8.5-Q16;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;c:\bin;C:\strawberry\c\bin;C:\strawberry\perl\site\bin;C:\strawberry\perl\bin;C:\Program
Files (x86)\PuTTY;C:\Program Files (x86)\Git\cmd;C:\Program Files
(x86)\Common Files\Adobe\AGL
    PERL_BADLANG (unset)
    SHELL (unset)


-- 
Jim Davis

```
</details>

**Comment by p5pRT** [comment]
### From @jkeenan

On Mon Sep 30 20&#8203;:09&#8203;:00 2013\, revhippie@&#8203;gmail\.com wrote&#8203;:
> This is a bug report for perl from revhippie@&#8203;gmail\.com\,
> generated with the help of perlbug 1\.39 running under perl 5\.16\.3\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> 'perldoc \-f splice' gives some example code&#8203;:
> 
>     sub aeq \{  \# compare two list values
>         my\(@&#8203;a\) = splice\(@&#8203;\_\,0\,shift\);
>         my\(@&#8203;b\) = splice\(@&#8203;\_\,0\,shift\);
>         return 0 unless @&#8203;a == @&#8203;b;  \# same len?
>         while \(@&#8203;a\) \{
>             return 0 if pop\(@&#8203;a\) ne pop\(@&#8203;b\);
>         \}
>         return 1;
>     \}
>     if \(&aeq\($len\,@&#8203;foo\[1\.\.$len\]\,0\+@&#8203;bar\,@&#8203;bar\)\) \{ \.\.\. \}
> 
> That looks like an unnecessary use of the ampersand\.  Mildly less\-
>    confusing
> replacement&#8203;:
> 
>     if \(aeq\($len\,@&#8203;foo\[1\.\.$len\]\,0\+@&#8203;bar\,@&#8203;bar\)\) \{ \.\.\. \}
> 
> Although\, to be honest\, that seems like a terribly convoluted example\.
> 

Most of the code in the subroutine example was contributed by a certain
Mr Wall on my mother's birthday in 1994 \(attached\)\.

Could you \(or anyone\) take a stab at saying what the example is supposed
to be exemplifying?

If we had a consensus on that\, we could write a better example\.

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
### From @jkeenan

a0d0e21e \(Larry Wall           1994\-10\-17 23&#8203;:00&#8203;:00 \+0000 5491\)     sub aeq \{  \# compare two list values
5a964f20 \(Tom Christiansen     1998\-06\-13 16&#8203;:19&#8203;:32 \-0600 5492\)        my\(@&#8203;a\) = splice\(@&#8203;\_\,0\,shift\);
5a964f20 \(Tom Christiansen     1998\-06\-13 16&#8203;:19&#8203;:32 \-0600 5493\)        my\(@&#8203;b\) = splice\(@&#8203;\_\,0\,shift\);
a0d0e21e \(Larry Wall           1994\-10\-17 23&#8203;:00&#8203;:00 \+0000 5494\)        return 0 unless @&#8203;a == @&#8203;b;       \# same len?
a0d0e21e \(Larry Wall           1994\-10\-17 23&#8203;:00&#8203;:00 \+0000 5495\)        while \(@&#8203;a\) \{
a0d0e21e \(Larry Wall           1994\-10\-17 23&#8203;:00&#8203;:00 \+0000 5496\)            return 0 if pop\(@&#8203;a\) ne pop\(@&#8203;b\);
a0d0e21e \(Larry Wall           1994\-10\-17 23&#8203;:00&#8203;:00 \+0000 5497\)        \}
a0d0e21e \(Larry Wall           1994\-10\-17 23&#8203;:00&#8203;:00 \+0000 5498\)        return 1;
a0d0e21e \(Larry Wall           1994\-10\-17 23&#8203;:00&#8203;:00 \+0000 5499\)     \}
a0d0e21e \(Larry Wall           1994\-10\-17 23&#8203;:00&#8203;:00 \+0000 5500\)     if \(&aeq\($len\,@&#8203;foo\[1\.\.$len\]\,0\+@&#8203;bar\,@&#8203;bar\)\) \{ \.\.\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @ikegami

On Wed\, Oct 2\, 2013 at 9&#8203;:30 PM\, James E Keenan via RT \<
perlbug\-followup@&#8203;perl\.org> wrote&#8203;:

> Could you \(or anyone\) take a stab at saying what the example is supposed
> to be exemplifying?
>

It's using splice \(and shift\) to split 3\, 'a'\, 'b'\, 'c'\, 4\, 'd'\, 'e'\, 'f'\,
'g' into two arrays\.

The rest of the code is just a reason to do that\.

If we had a consensus on that\, we could write a better example\.
>

An n\-at\-a\-time algorithm?

while \(@&#8203;\_\) \{
&nbsp;    my  \($key\, $val\) = splice\(@&#8203;\_\, 0\, 2\);
&nbsp;    \.\.\.
\}

**Comment by p5pRT** [comment]
### From @nwc10

On Wed\, Oct 02\, 2013 at 10&#8203;:53&#8203;:28PM \-0400\, Eric Brine wrote&#8203;:
> On Wed\, Oct 2\, 2013 at 9&#8203;:30 PM\, James E Keenan via RT \<
> perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> 
> > Could you \(or anyone\) take a stab at saying what the example is supposed
> > to be exemplifying?
> >
> 
> It's using splice \(and shift\) to split 3\, 'a'\, 'b'\, 'c'\, 4\, 'd'\, 'e'\, 'f'\,
> 'g' into two arrays\.
> 
> The rest of the code is just a reason to do that\.

Yes\, and it's not an idiomatic way to write the code\. It would be much
clearer to pass two references\.

> If we had a consensus on that\, we could write a better example\.
> >
> 
> An n\-at\-a\-time algorithm?
> 
> while \(@&#8203;\_\) \{
>     my  \($key\, $val\) = splice\(@&#8203;\_\, 0\, 2\);
>     \.\.\.
> \}

n\-at\-a\-time is the usual reason I end up using splice\. I think that an
example using n\-at\-time would be much better\. I'd also suggest avoiding
using the names $key\, $val\. Possibly 3\-at\-a\-time would be least ambiguous\,
as a 3\-at\-a\-time list can't have come from flattening a hash\, there isn't
another obvious way to implement it\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From @xdg

On Fri\, Oct 4\, 2013 at 3&#8203;:22 AM\, Nicholas Clark \<nick@&#8203;ccl4\.org> wrote&#8203;:
>> An n\-at\-a\-time algorithm?
>>
>> while \(@&#8203;\_\) \{
>>     my  \($key\, $val\) = splice\(@&#8203;\_\, 0\, 2\);
>>     \.\.\.
>> \}
>
> n\-at\-a\-time is the usual reason I end up using splice\. I think that an
> example using n\-at\-time would be much better\. I'd also suggest avoiding
> using the names $key\, $val\. Possibly 3\-at\-a\-time would be least ambiguous\,
> as a 3\-at\-a\-time list can't have come from flattening a hash\, there isn't
> another obvious way to implement it\.

I agree\.  Either 3\-at\-a\-time or fully genericize it to N&#8203;:

&nbsp;    my $n = shift;
&nbsp;    while \(@&#8203;\_\) \{
&nbsp;        my @&#8203;part = splice\(@&#8203;\_\, 0\, $n\);
&nbsp;        \.\.\.\.
&nbsp;    \}

\-\- 
David Golden \<xdg@&#8203;xdg\.me>
Take back your inbox\! → http&#8203;://www\.bunchmail\.com/
Twitter/IRC&#8203;: @&#8203;xdg

**Comment by p5pRT** [comment]
### From @rjbs

\* David Golden \<xdg@&#8203;xdg\.me> \[2013\-10\-04T10&#8203;:04&#8203;:20\]
> I agree\.  Either 3\-at\-a\-time or fully genericize it to N&#8203;:

&nbsp;  commit 498b759bd12b378c541fc401200803b8c88c4a9b
&nbsp;  Author&#8203;: Ricardo Signes \<rjbs@&#8203;cpan\.org>
&nbsp;  Date&#8203;:   Fri Oct 4 10&#8203;:42&#8203;:49 2013 \-0400

&nbsp;      perlfunc&#8203;: replace splice example with simpler one

&nbsp;      based on suggestions from Nicholas Clark and David Golden on
&nbsp;      perl5\-porters

New example&#8203;:

&nbsp;  C\<splice> can be used\, for example\, to implement n\-ary queue processing&#8203;:

&nbsp;      sub nary\_print \{
&nbsp;        my $n = shift;
&nbsp;        while \(my @&#8203;next\_n = splice @&#8203;\_\, 0\, $n\) \{
&nbsp;          say join q\{ \-\- \}\, @&#8203;next\_n;
&nbsp;        \}
&nbsp;      \}

&nbsp;      nary\_print\(3\, qw\(a b c d e f g h\)\);
&nbsp;      \# prints&#8203;:
&nbsp;      \#   a \-\- b \-\- c
&nbsp;      \#   d \-\- e \-\- f
&nbsp;      \#   g \-\- h

\-\- 
rjbs

**Comment by p5pRT** [comment]
### From @rjbs

Resolved by 498b759bd12b378\.  Please let me know if there are significant problems\.

\-\- 
rjbs

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Resolved by 498b759bd12b378\.  Please let me know if there are significant problems\.

\-\- 
rjbs

**Comment by p5pRT** [comment]
@rjbs - Status changed from 'open' to 'resolved'
